File: src/commands/rule.ts
import type { Bot } from "grammy";
import { getGroupConfig, updateGroupRule } from "../db/queries";
import { checkAdminPermission } from "../utils/permissions";
import { replyOptions } from "../utils/reply";

export function registerRuleCommands(bot: Bot, db: D1Database): void {
	bot.command("rule", async (ctx) => {
		const { allowed, groupId } = await checkAdminPermission(db, ctx);
		if (!allowed || !groupId) {
			await ctx.reply("只有管理员才能使用此命令。", replyOptions(ctx));
			return;
		}

		const rule = ctx.match?.toString().trim();

		if (!rule) {
			const config = await getGroupConfig(db, groupId);
			if (config?.rule) {
				await ctx.reply(
					`当前群规:\n\n${config.rule}\n\n使用 /rule <内容> 修改群规\n使用 /rule off 清除群规`,
					replyOptions(ctx),
				);
			} else {
				await ctx.reply(
					"当前未设置群规。\n\n使用 /rule <内容> 设置群规",
					replyOptions(ctx),
				);
			}
			return;
		}

		if (rule === "off") {
			await updateGroupRule(db, groupId, "");
			await ctx.reply("✅ 群规已清除。", replyOptions(ctx));
			return;
		}

		await updateGroupRule(db, groupId, rule);
		await ctx.reply(
			"✅ 群规已设置。入群认证时将强制阅读群规。",
			replyOptions(ctx),
		);
	});
}

File: internal/handler/rule.go
package handler

import (
	"context"
	"log/slog"

	"github.com/ChouChiu/neptune/internal/db"
	"github.com/ChouChiu/neptune/internal/util"
	tgbot "github.com/go-telegram/bot"
	"github.com/go-telegram/bot/models"
)

// Rule returns a handler for the /rule command.
func Rule(database *db.DB) tgbot.HandlerFunc {
	return func(ctx context.Context, b *tgbot.Bot, update *models.Update) {
		allowed, groupID := requireAdmin(ctx, b, database, update)
		if !allowed {
			return
		}

		rule := ""
		if update.Message != nil && update.Message.Text != "" {
			rule = commandArgs(update.Message.Text, "/rule")
		}
		rule = trimSpace(rule)

		if rule == "" {
			config, err := database.GetGroupConfig(groupID)
			if err != nil {
				slog.Error("Failed to get group config for rule", "error", err)
				return
			}

			if config != nil && config.Rule != "" {
				sendMessage(ctx, b, &tgbot.SendMessageParams{
					ChatID:          update.Message.Chat.ID,
					Text:            "当前群规:\n\n" + config.Rule + "\n\n使用 /rule <内容> 修改群规\n使用 /rule off 清除群规",
					ReplyParameters: util.ReplyOptions(update.Message),
				})
			} else {
				sendMessage(ctx, b, &tgbot.SendMessageParams{
					ChatID:          update.Message.Chat.ID,
					Text:            "当前未设置群规。\n\n使用 /rule <内容> 设置群规",
					ReplyParameters: util.ReplyOptions(update.Message),
				})
			}
			return
		}

		if rule == "off" {
			if err := database.UpdateGroupRule(groupID, ""); err != nil {
				slog.Error("Failed to clear rule", "error", err)
				return
			}
			sendMessage(ctx, b, &tgbot.SendMessageParams{
				ChatID:          update.Message.Chat.ID,
				Text:            "✅ 群规已清除。",
				ReplyParameters: util.ReplyOptions(update.Message),
			})
			return
		}

		if len(rule) > 2048 {
			sendMessage(ctx, b, &tgbot.SendMessageParams{
				ChatID:          update.Message.Chat.ID,
				Text:            "群规过长（最大 2048 字符）。",
				ReplyParameters: util.ReplyOptions(update.Message),
			})
			return
		}

		if err := database.UpdateGroupRule(groupID, rule); err != nil {
			slog.Error("Failed to update rule", "error", err)
			return
		}

		sendMessage(ctx, b, &tgbot.SendMessageParams{
			ChatID:          update.Message.Chat.ID,
			Text:            "✅ 群规已设置。入群认证时将强制阅读群规。",
			ReplyParameters: util.ReplyOptions(update.Message),
		})
	}
}
